File: backend-springboot/src/main/java/com/splitwiseplus/controller/UserController.java
package com.splitwiseplus.controller;

import com.splitwiseplus.model.User;
import com.splitwiseplus.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import java.util.List;

@RestController
@RequestMapping("/api/users")
@CrossOrigin(origins = "*")
public class UserController {
    @Autowired
    private UserRepository userRepository;

    @Autowired
    private org.springframework.security.authentication.AuthenticationManager authenticationManager;

    @Autowired
    private com.splitwiseplus.security.JwtUtils jwtUtils;

    @Autowired
    private org.springframework.security.crypto.password.PasswordEncoder encoder;

    @PostMapping("/login")
    public org.springframework.http.ResponseEntity<?> authenticateUser(@RequestBody java.util.Map<String, String> loginRequest) {
        try {
            org.springframework.security.core.Authentication authentication = authenticationManager.authenticate(
                    new org.springframework.security.authentication.UsernamePasswordAuthenticationToken(loginRequest.get("email"), loginRequest.get("password")));

            org.springframework.security.core.context.SecurityContextHolder.getContext().setAuthentication(authentication);
            String jwt = jwtUtils.generateJwtToken(loginRequest.get("email"));
            
            com.splitwiseplus.security.UserDetailsImpl userDetails = (com.splitwiseplus.security.UserDetailsImpl) authentication.getPrincipal();
            
            java.util.Map<String, Object> response = new java.util.HashMap<>();
            response.put("token", jwt);
            response.put("_id", userDetails.getId());
            response.put("name", userDetails.getName());
            response.put("email", userDetails.getUsername());
            response.put("role", userDetails.getRole());
            
            return org.springframework.http.ResponseEntity.ok(response);
        } catch (org.springframework.security.core.AuthenticationException e) {
            return org.springframework.http.ResponseEntity.status(401).body(java.util.Map.of("message", "Invalid Credentials"));
        }
    }

    @PostMapping("/register")
    public org.springframework.http.ResponseEntity<?> registerUser(@RequestBody User user) {
        if (userRepository.findByEmail(user.getEmail()).isPresent()) {
            return org.springframework.http.ResponseEntity.badRequest().body(java.util.Map.of("message", "User already exists"));
        }
        user.setPassword(encoder.encode(user.getPassword()));
        userRepository.save(user);
        return org.springframework.http.ResponseEntity.ok(java.util.Map.of("message", "User registered successfully!"));
    }

    @GetMapping("/")
    public List<User> getUsers() {
        return userRepository.findAll();
    }

    @PostMapping("/forgot-password")
    public org.springframework.http.ResponseEntity<?> forgotPassword(@RequestBody java.util.Map<String, String> request) {
        String email = request.get("email");
        User user = userRepository.findByEmail(email).orElse(null);
        if (user != null) {
            String token = java.util.UUID.randomUUID().toString();
            user.setResetPasswordToken(token);
            user.setResetPasswordExpire(new java.util.Date(System.currentTimeMillis() + 3600000)); // 1 hour
            userRepository.save(user);
            
            java.util.Map<String, String> response = new java.util.HashMap<>();
            response.put("message", "Reset link generated");
            response.put("resetToken", token);
            return org.springframework.http.ResponseEntity.ok(response);
        }
        return org.springframework.http.ResponseEntity.ok(java.util.Map.of("message", "Reset link sent if email exists"));
    }

    @PostMapping("/reset-password")
    public org.springframework.http.ResponseEntity<?> resetPassword(@RequestBody java.util.Map<String, String> request) {
        String token = request.get("token");
        String password = request.get("password");
        
        User user = userRepository.findByResetPasswordToken(token).orElse(null);
        if (user == null) {
            return org.springframework.http.ResponseEntity.badRequest().body(java.util.Map.of("message", "Invalid token"));
        }
        
        if (user.getResetPasswordExpire() != null && user.getResetPasswordExpire().before(new java.util.Date())) {
            return org.springframework.http.ResponseEntity.badRequest().body(java.util.Map.of("message", "Token expired"));
        }
        
        user.setPassword(encoder.encode(password));
        user.setResetPasswordToken(null);
        user.setResetPasswordExpire(null);
        userRepository.save(user);
        
        return org.springframework.http.ResponseEntity.ok(java.util.Map.of("message", "Password reset successful"));
    }

    @SuppressWarnings("null")
    @GetMapping("/{id}")
    public User getUserById(@PathVariable String id) {
        return userRepository.findById(id).orElse(null);
    }

    @PutMapping("/update-profile")
    public User updateProfile(@RequestBody User userDetails) {
        // In a real app, get user from security context
        User user = userRepository.findByEmail(userDetails.getEmail()).orElse(null);
        if (user != null) {
            user.setName(userDetails.getName());
            return userRepository.save(user);
        }
        return null;
    }
}

File: backend/controllers/userController.js
const User = require("../models/User")
const bcrypt = require("bcryptjs")
const jwt = require("jsonwebtoken")
const crypto = require("crypto")

// REGISTER
exports.registerUser = async (req, res) => {
  try {
    const { name, email, password } = req.body

    const existingUser = await User.findOne({ email })
    if (existingUser) {
      return res.status(400).json({ message: "User already exists" })
    }

    // hash password
    const salt = await bcrypt.genSalt(10)
    const hashedPassword = await bcrypt.hash(password, salt)

    const user = await User.create({ name, email, password: hashedPassword })

    // create token
    const token = jwt.sign({ id: user._id }, process.env.JWT_SECRET || "secret", {
      expiresIn: "7d",
    })

    res.status(201).json({ _id: user._id, name: user.name, email: user.email, role: user.role, token })
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}

// LOGIN
exports.loginUser = async (req, res) => {
  try {
    const { email, password } = req.body

    const user = await User.findOne({ email })

    if (!user) {
      return res.status(401).json({ message: "Invalid Credentials" })
    }

    const isMatch = await bcrypt.compare(password, user.password)
    if (!isMatch) {
      // fallback: handle legacy plain-text passwords (migrate to hashed)
      if (user.password === password) {
        try {
          const salt = await bcrypt.genSalt(10)
          user.password = await bcrypt.hash(password, salt)
          await user.save()
        } catch (e) {
          console.error('Password migration failed', e)
        }
      } else {
        return res.status(401).json({ message: "Invalid Credentials" })
      }
    }

    const token = jwt.sign({ id: user._id }, process.env.JWT_SECRET || "secret", {
      expiresIn: "7d",
    })

    res.json({ _id: user._id, name: user.name, email: user.email, role: user.role, token })
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}

// GET ALL USERS
exports.getUsers = async (req, res) => {
  try {
    const users = await User.find().select("-password")
    res.json(users)
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}

// FORGOT PASSWORD
exports.forgotPassword = async (req, res) => {
  try {
    const { email } = req.body
    const user = await User.findOne({ email })

    if (!user) {
      return res.status(404).json({ message: "User not found" })
    }

    // Generate reset token
    const resetToken = crypto.randomBytes(20).toString("hex")
    user.resetPasswordToken = crypto.createHash("sha256").update(resetToken).digest("hex")
    user.resetPasswordExpire = Date.now() + 3600000 // 1 hour

    await user.save()

    // In a real app, send email. Here, we return it for demo purposes.
    res.json({ message: "Token generated", resetToken })
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}

// RESET PASSWORD
exports.resetPassword = async (req, res) => {
  try {
    const { token, password } = req.body
    const hashedToken = crypto.createHash("sha256").update(token).digest("hex")

    const user = await User.findOne({
      resetPasswordToken: hashedToken,
      resetPasswordExpire: { $gt: Date.now() },
    })

    if (!user) {
      return res.status(400).json({ message: "Invalid or expired token" })
    }

    // Set new password
    const salt = await bcrypt.genSalt(10)
    user.password = await bcrypt.hash(password, salt)
    user.resetPasswordToken = undefined
    user.resetPasswordExpire = undefined

    await user.save()

    res.json({ message: "Password reset successful" })
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}

// UPDATE PROFILE
exports.updateProfile = async (req, res) => {
  try {
    const { id, name, email } = req.body
    const user = await User.findById(id)

    if (!user) return res.status(404).json({ message: "User not found" })

    if (name) user.name = name
    if (email) user.email = email

    await user.save()
    res.json({ _id: user._id, name: user.name, email: user.email, role: user.role })
  } catch (err) {
    res.status(500).json({ message: err.message })
  }
}
